kernel/src/process.rs

```diff
@@ -119,127 +119,137 @@ impl fmt::Debug for ProcessLoadError {
-            // entry is not a process.
-            let process_memory_length = if header_length > 0 {
-                // Try to create a process object from that app slice.
-                let (process, process_memory_length) = Process::create(
+        // Now we can get a slice which only encompasses the length of flash
+        // described by this tbf header.  We will either parse this as an actual
+        // app, or skip over this region.
+        let entry_flash = remaining_flash
+            .get(0..entry_length as usize)
+            .ok_or(ProcessLoadError::NotEnoughFlash)?;
+
+        // Need to reassign remaining_memory in every iteration so the compiler
+        // knows it will not be re-borrowed.
+        remaining_memory = if header_length > 0 {
+            // If we found an actual app header, try to create a `Process`
+            // object. We also need to shrink the amount of remaining memory
+            // based on whatever is assigned to the new process if one is
+            // created.
+
+            // Try to create a process object from that app slice. If we don't
+            // get a process and we didn't get a loading error (aka we got to
+            // this point), then the app is a disabled process or just padding.
+            let (process_option, unused_memory) = unsafe {
+                Process::create(
                     kernel,
                     chip,
                     entry_flash,
                     header_length as usize,
                     version,
-                    app_memory_ptr,
-                    app_memory_size,
+                    remaining_memory,
                     fault_response,
                     i,
-                )?;
-
-                // Check to see if actually got a valid process to execute. If we
-                // didn't and we didn't get a loading error (aka we got to this
-                // point), then the app is a disabled process or just padding.
-                if process.is_some() {
-                    if config::CONFIG.debug_load_processes {
-                        debug!(
-                            "Loaded process[{}] from flash=[{:#010X}:{:#010X}] into sram=[{:#010X}:{:#010X}] = {:?}",
-                            i,
-                            entry_flash.as_ptr() as usize,
-                            entry_flash.as_ptr() as usize + entry_flash.len(),
-                            app_memory_ptr as usize,
-                            app_memory_ptr as usize + process_memory_length,
-                            process.map(|p| p.get_process_name())
-                        );
-                    }
-                    procs[i] = process;
-                    process_memory_length
-                } else {
-                    0
-                }
-            } else {
-                // We are just skipping over this region in flash.
-                0
+                )?
```

Comment:
I'm not sure we want to use the `?` operator here, as a non-fatal in one process (e.g. `MemoryAddressMismatch`) will stop later processes from loading.

Reply:
But if we don't, then the error doesn't get propagated to main.rs and the user won't know that the app wasn't loaded, putting us right back where we were.

Reply:
Mhm.., makes sense. I think in the long run, we'll probably want the process load policy to be configurable by the board, but that's an enhancement beyond this PR.